### ISSUE TITLE
Error: Unexpected end of JSON input when using empty package.json

### ISSUE THREAD

**Issue by vladshcherbin** [eslint/eslint#12568]
**The version of ESLint you are using.**
6.6.0

**The problem you want to solve.**
When I have an empty `package.json` file in nested folder, near checked files, eslint gives error:

```console
Error: Unexpected end of JSON input
```

There were similar issues previously - https://github.com/eslint/eslint/issues/11026#issuecomment-433744443, https://github.com/eslint/eslint/issues/7748 and there was no error in 5.16.0. Since 6.0.0 an error is thrown.

**Your take on the correct solution to problem.**
It would be nice to have an option to disable this error. I tried to ignore this json file, but didn't succeed.

This is useful for me because I want this file to be invalid on purpose so later I can check it.

**Are you willing to submit a pull request to implement this change?**
Probably not, no free time to dive in atm.

**Comment by platinumazure** [comment]
Hi @vladshcherbin, thanks for the issue.

I wonder if that message is being thrown while searching for an ESLint configuration file (or `eslintConfig` key in package.json)?

Do you have an ESLint configuration file? If not, you should create one by running ESLint with the `--init` comment line option. That could be a workaround to avoid this error being thrown. (Of course, I could be wrong and it could be something else.)

If you have a stack trace with this error, could you please post it? Thanks!

**Comment by kaicataldo** [comment]
Could you also explain what you mean by "empty `package.json`"? Does the file literally have nothing in it or does it have `{}`?

**Comment by vladshcherbin** [comment]
so, in my situation I have a similar structure:

```
|-- tests
|   |-- bad-json
|   |   |-- index.js
|   |   `-- package.json
|   `-- other-tests
|-- .eslintrc.js
`--  package.json
```

As you can see, here I have root `package.json` and `.eslintrc.js` config, both are valid and working. In `tests/bad-json` folder I have another `package.json` file, which is completely empty (not even a `{}`).

When I try to lint `tests` folder, I believe eslint tried to read this nested `tests/bad-json/package.json` file to find eslint configuration and gives me an error.

I tried to add this `package.json` file to `.eslintignore`, `override.excludedFiles`, but didn't have any luck. I got this error after updating from eslint v5 to v6.

<details>
 <summary>Error screenshot</summary>

![image](https://user-images.githubusercontent.com/6711845/68931978-b91cf300-07a2-11ea-9d7d-d553b703940b.png)

</details>

**Comment by kaicataldo** [comment]
Thanks for the info! This will make it much easier to debug.

**Comment by mysticatea** [comment]
> When I try to lint tests folder, I believe eslint tries to read this nested tests/bad-json/package.json file to find eslint configuration and gives me an error.

You are right. In that case, you need `/tests/bad-json` in your `.eslintignore`, or a valid ESLint configuration file that has [a higher priority than `package.json`](https://eslint.org/docs/user-guide/configuring#configuration-file-formats) such as `/tests/bad-json/.eslintrc.yml`.

Because ESLint reads `/tests/bad-json/package.json` when it entered `/tests/bad-json` directory in order to check if the config file change how ESLint traverses the subdirectory via RFC20 or RFC22. (so `/tests/bad-json/package.json` can have a setting that ignores/unignores `/tests/bad-json/*`.)

**Comment by kaicataldo** [comment]
@mysticatea Do you think it would make sense to ignore empty package.json files, or at least only log a warning but not exit with an error code? Seems like if the package.json file doesn’t have any contents that it clearly is not intended to be used to configure ESLint.

**Comment by mysticatea** [comment]
I feel odd if ESLint handles empty config files in special. A `tests/bad-json/package.json` may contain invalid JSON. So this is how ESLint handles parsing errors in config files.

The current strategy is to stop at a parsing error in config files. To print a warning and continue sounds good, but I have a concern about arbitrary outputs from outside of configured formatter.

**Comment by mysticatea** [comment]
Honestly, I think reasonable if we going to deprecate config files in subdirectories.

**Comment by vladshcherbin** [comment]
> You are right. In that case, you need `/tests/bad-json` in your `.eslintignore`, or a valid ESLint configuration file that has [a higher priority than `package.json`](https://eslint.org/docs/user-guide/configuring#configuration-file-formats) such as `/tests/bad-json/.eslintrc.yml`.

I tried different variants of adding `/tests/bad-json` to `.eslintignore` and only 1 variant worked - when I ignore the whole directory. However, I need files inside of this folder to be linted too, for example `tests/bad-json/index.js`. Some of my tries:

```yml
# ignores folder contents
/tests/bad-json

# json error (same with ** or *)
/tests/bad-json/**

# ignores folder contents
/tests/bad-json
!/tests/bad-json/index.js
```

So, I believe current workaround in my situation is to have a higher priority config file in this folder.

**Comment by platinumazure** [comment]
I think we should treat invalid or empty object package.json the same way as we treat lack of configuration files of any other type. The only thing in package.json that counts as an ESLint configuration is an `eslintConfig` key and object. Anything else should not matter to ESLint.

I think we should throw if the `eslintConfig` value is invalid for some reason (e.g., it's an array or string rather than an object). But, if we can't even see an `eslintConfig` key, then there is no configuration for ESLint to look at and we know this for a fact. Throwing on invalid JSON (for package.json) is unnecessarily intrusive, in my opinion.

**Comment by mysticatea** [comment]
If parsing errors happened, it doesn't know the errors are in `eslintConfig` or not. This is similar to that core rules cannot work if parsing errors happened.

Hmm, how about the following steps:

1. Read the content of `package.json`
2. Check if the content includes the constant string `"eslintConfig"` or not. If not, skip parsing.
3. Parse the content as JSON.

Therefore, if the package.json file doesn't contain ESLint config, ESLint skips the package.json file.

**Comment by kaicataldo** [comment]
> If parsing errors happened, it doesn't know the errors are in eslintConfig or not.

This is a good point. As I think on this more, I'm of the mind that invalid JSON should still throw, but that if we wanted to ignore completely empty JSON files, that seems like reasonable behavior.  As I think about this more, I actually think the new behavior in 6.0.0 makes a lot of sense.

@vladshcherbin Is there a particular reason the files can't just contain `{}` to indicate they are empty? Keeping this behavior because some users might be relying on an error being thrown as a TODO/reminder feels out of scope of the tool's area of responsibility.

**Comment by vladshcherbin** [comment]
@vladshcherbin in a plugin I have user can set input `package.json` file path, if this file has invalid json format, I throw an error. In tests I want to test this case by providing invalid json file myself to see if correct error is thrown.

**Comment by kaicataldo** [comment]
So, to clarify, you want this feature for testing purposes? If so, there are many assertion libraries that include built-in support for testing this, or you can use try/catch or Promises and assert in the catch block or in the rejection callback as well.

**Comment by vladshcherbin** [comment]
@kaelzhang no, I use eslint only for linting, nothing else. When I try to lint `tests` folder which contains this empty package.json (created for tests) eslint throws an error when sees it and tries to use.

**Comment by vladshcherbin** [comment]
@kaicataldo here is [test repo for my case](https://github.com/vladshcherbin/eslint-invalid-json).

In `src` folder function [throws if passed package.json](https://github.com/vladshcherbin/eslint-invalid-json/blob/master/src/index.js#L3-L9) is invalid.
In `tests` folder there is [test for this error](https://github.com/vladshcherbin/eslint-invalid-json/blob/master/tests/index.test.js#L4).
In `tests/bad-json` there is this empty `package.json` file, used in this test.


When I try to lint `src` and `tests` folders, error is thrown. Hope this clears things up now :)

**Comment by kaicataldo** [comment]
Appreciate all the info! I think the suggestion [here](https://github.com/eslint/eslint/issues/12568#issuecomment-554423743) is the right solution to this. As someone who works on language tooling, I’ve encountered these kinds of issues a lot, and what I tend to see done is to put test fixture files in a ‘test/fixtures` directory and ignore the whole directory when linting.
